File: Chatbots/Python/Source/Scoring/tfidf.py
from __future__ import annotations

from typing import Dict, List, Tuple
from math import sqrt

from ..dataModel import (
    Card,
    TfidfIndex,
    AnswerHit,
)
from ..config import ParserConfig
from ..normalise import normalise_for_matching
from ..tokenise import tokenise


def build_tfidf_index(
    candidate_cards: List[Card],
    stopwords: set[str],
    parser_config: ParserConfig,
) -> TfidfIndex:
    """
    Build a TF-IDF index for the given candidate cards.
    """
    documents: List[Card] = list(candidate_cards)
    document_count = len(documents)

    term_frequencies_per_document: List[Dict[str, int]] = []
    document_frequency: Dict[str, int] = {}

    for card in documents:
        tokens = tokenise(card.question_text, stopwords, parser_config)
        frequency_map: Dict[str, int] = {}
        for token in tokens:
            frequency_map[token] = frequency_map.get(token, 0) + 1
        term_frequencies_per_document.append(frequency_map)

        seen_terms = set(frequency_map.keys())
        for term in seen_terms:
            document_frequency[term] = document_frequency.get(term, 0) + 1

    idf_map: Dict[str, float] = {}
    for term, df in document_frequency.items():
        from math import log

        if parser_config.idf_smoothing:
            idf_value = log((document_count + 1.0) / (df + 1.0)) + 1.0
        else:
            if df == 0:
                continue
            idf_value = log(document_count / df)
        idf_map[term] = idf_value

    inverted_index: Dict[str, List[Tuple[int, int]]] = {}
    for doc_id, frequency_map in enumerate(term_frequencies_per_document):
        for term, raw_tf in frequency_map.items():
            inverted_index.setdefault(term, []).append((doc_id, raw_tf))

    document_norms: List[float] = [0.0] * document_count
    document_token_counts: List[int] = [card.question_token_count for card in documents]

    for doc_id, frequency_map in enumerate(term_frequencies_per_document):
        sum_of_squares = 0.0
        for term, raw_tf in frequency_map.items():
            idf_value = idf_map.get(term, 0.0)
            weight = raw_tf * idf_value
            sum_of_squares += weight * weight
        document_norms[doc_id] = sqrt(sum_of_squares) if sum_of_squares > 0.0 else 0.0

    return TfidfIndex(
        documents=documents,
        inverted_index=inverted_index,
        idf=idf_map,
        document_norms=document_norms,
        document_token_counts=document_token_counts,
    )


def score_tfidf(
    query_text: str,
    index: TfidfIndex,
    stopwords: set[str],
    parser_config: ParserConfig,
    top_k: int = 1,
) -> List[AnswerHit]:
    """
    Score documents using cosine similarity between query and document TF-IDF vectors.
    """
    if top_k < 1:
        top_k = 1
    if not index.documents:
        return []
    
    normalised_query = normalise_for_matching(query_text, parser_config)
    query_tokens = tokenise(normalised_query, stopwords, parser_config)
    if not query_tokens:
        return []

    query_term_frequency: Dict[str, int] = {}
    for token in query_tokens:
        query_term_frequency[token] = query_term_frequency.get(token, 0) + 1

    query_weights: Dict[str, float] = {}
    sum_of_squares_query = 0.0
    for term, raw_tf in query_term_frequency.items():
        idf_value = index.idf.get(term)
        if idf_value is None:
            continue
        weight = raw_tf * idf_value
        query_weights[term] = weight
        sum_of_squares_query += weight * weight
    query_norm = sqrt(sum_of_squares_query) if sum_of_squares_query > 0.0 else 0.0
    if query_norm == 0.0:
        return []

    document_dot: Dict[int, float] = {}
    document_overlap_count: Dict[int, int] = {}

    for term, query_weight in query_weights.items():
        postings = index.inverted_index.get(term)
        if not postings:
            continue
        idf_value = index.idf.get(term, 0.0)
        for doc_id, raw_tf_in_document in postings:
            document_dot[doc_id] = document_dot.get(doc_id, 0.0) + (query_weight * (raw_tf_in_document * idf_value))
            document_overlap_count[doc_id] = document_overlap_count.get(doc_id, 0) + 1

    if not document_dot:
        return []

    scored_rows: List[Tuple[float, int, int, str]] = []
    for doc_id, dot_value in document_dot.items():
        document_norm = index.document_norms[doc_id]
        if document_norm == 0.0:
            continue
        cosine = dot_value / (document_norm * query_norm)
        overlap = document_overlap_count.get(doc_id, 0)
        question_token_count = index.document_token_counts[doc_id]
        guid = index.documents[doc_id].guid
        scored_rows.append((cosine, overlap, question_token_count, guid))

    scored_rows.sort(key=lambda row: (-row[0], -row[1], row[2], row[3]))

    hits: List[AnswerHit] = []
    for cosine, overlap, question_token_count, guid in scored_rows[:top_k]:
        card = _lookup_card_by_guid(index, guid)
        hits.append(
            AnswerHit(
                guid=guid,
                score=float(cosine),
                deck_path=card.deck_path,
                question_preview=_short_preview(card.question_text),
            )
        )
    return hits


def _lookup_card_by_guid(index: TfidfIndex, guid: str) -> Card:
    for card in index.documents:
        if card.guid == guid:
            return card
    raise KeyError(f"GUID not found in index: {guid}")


def _short_preview(text: str, max_length: int = 120) -> str:
    trimmed = text.strip()
    if len(trimmed) <= max_length:
        return trimmed
    return trimmed[: max_length - 1] + "…"

File: Chatbots/Rust/Source/src/scoring/tfidf.rs
use std::collections::HashMap;
use std::collections::HashSet;

use crate::config::ParserConfig;
use crate::data_model::{make_hit, AnswerHit, Card};
use crate::tokenise::{tokenise, tokenise_to_set};

/// A single TF vector for a document (question), with metadata for display.
#[derive(Debug, Clone)]
pub struct DocumentEntry {
    pub guid: String,
    pub deck_path: Vec<String>,
    pub question_preview: String,
    pub term_frequencies: HashMap<String, f32>,
    pub token_count: usize,
}

/// The TF–IDF index with per-document TF maps, global IDF weights, and precomputed norms.
#[allow(dead_code)]
#[derive(Debug, Clone)]
pub struct TfidfIndex {
    pub documents: Vec<DocumentEntry>,
    pub inverse_document_frequency: HashMap<String, f32>,
    pub document_l2_norms: Vec<f32>,
    pub vocabulary_size: usize,
    pub document_count: usize,
}

/// Build a TF–IDF index over the candidate cards (question text only).
/// - TF = raw term count
/// - IDF = log((N + 1) / (df + 1)) + 1
/// - Document vectors L2-normalised during scoring via precomputed norms
pub fn build_tfidf_index(
    candidate_cards: &[Card],
    stopword_set: &HashSet<String>,
    parser_config: &ParserConfig,
) -> TfidfIndex {
    let mut document_entries: Vec<DocumentEntry> = Vec::with_capacity(candidate_cards.len());
    let mut document_frequency_map: HashMap<String, usize> = HashMap::new();

    // 1) Build documents with raw TF and gather DF
    for card in candidate_cards {
        let token_vector = tokenise(&card.question_text, stopword_set, parser_config);
        let token_count = token_vector.len();

        let mut term_counts: HashMap<String, usize> = HashMap::new();
        for token_text in token_vector {
            *term_counts.entry(token_text).or_insert(0) += 1;
        }

        // Increment DF once per unique term in this document
        let unique_terms = term_counts.keys().cloned().collect::<HashSet<String>>();
        for unique_term in unique_terms {
            *document_frequency_map.entry(unique_term).or_insert(0) += 1;
        }

        // Convert usize counts to f32 early for speed later
        let term_frequencies = term_counts
            .into_iter()
            .map(|(term_text, count_value)| (term_text, count_value as f32))
            .collect::<HashMap<String, f32>>();

        let document_entry = DocumentEntry {
            guid: card.guid.clone(),
            deck_path: card.deck_path.clone(),
            question_preview: card.question_text.clone(),
            term_frequencies,
            token_count,
        };
        document_entries.push(document_entry);
    }

    let document_count = document_entries.len();
    let mut inverse_document_frequency: HashMap<String, f32> = HashMap::new();

    // 2) Compute IDF with smoothing: log((N + 1) / (df + 1)) + 1
    for (term_text, document_frequency) in document_frequency_map.into_iter() {
        let numerator = (document_count as f32) + 1.0;
        let denominator = (document_frequency as f32) + 1.0;
        let idf_value = (numerator / denominator).ln() + 1.0;
        inverse_document_frequency.insert(term_text, idf_value);
    }

    // 3) Precompute document vector norms (L2)
    let mut document_l2_norms: Vec<f32> = Vec::with_capacity(document_entries.len());
    for document_entry in &document_entries {
        let mut squared_sum: f32 = 0.0;
        for (term_text, term_frequency) in &document_entry.term_frequencies {
            let idf_value = *inverse_document_frequency.get(term_text).unwrap_or(&0.0);
            let weighted_value = (*term_frequency) * idf_value;
            squared_sum += weighted_value * weighted_value;
        }
        document_l2_norms.push(squared_sum.sqrt());
    }

    let vocabulary_size = inverse_document_frequency.len();

    TfidfIndex {
        documents: document_entries,
        inverse_document_frequency,
        document_l2_norms,
        vocabulary_size,
        document_count,
    }
}

/// Rank candidates by cosine similarity between the query TF–IDF vector and each document.
/// Tie-breakers (after equal similarity):
/// 1) Greater count of overlapping non-stopword tokens
/// 2) Shorter candidate question by token count
/// 3) Lexicographic order of GUID
pub fn score_tfidf(
    query_text: &str,
    tfidf_index: &TfidfIndex,
    stopword_set: &HashSet<String>,
    parser_config: &ParserConfig,
    top_k: usize,
) -> Vec<AnswerHit> {
    // 1) Tokenise the query and build its TF map
    let query_token_vector = tokenise(query_text, stopword_set, parser_config);
    if query_token_vector.is_empty() {
        return Vec::new();
    }

    let mut query_term_counts: HashMap<String, usize> = HashMap::new();
    for token_text in query_token_vector.iter() {
        *query_term_counts.entry(token_text.clone()).or_insert(0) += 1;
    }

    // 2) Convert to TF–IDF and compute query norm
    let mut query_weighted_map: HashMap<String, f32> = HashMap::new();
    for (term_text, count_value) in &query_term_counts {
        let idf_value = *tfidf_index
            .inverse_document_frequency
            .get(term_text)
            .unwrap_or(&0.0);
        if idf_value == 0.0 {
            continue; // term unseen in the corpus → contributes nothing
        }
        query_weighted_map.insert(term_text.clone(), (*count_value as f32) * idf_value);
    }

    let mut query_squared_sum: f32 = 0.0;
    for weighted_value in query_weighted_map.values() {
        query_squared_sum += weighted_value * weighted_value;
    }
    let query_l2_norm = query_squared_sum.sqrt();
    if query_l2_norm == 0.0 {
        return Vec::new();
    }

    // Prepare set for overlap-based tie-breaker
    let query_token_set = tokenise_to_set(query_text, stopword_set, parser_config);

    // 3) Score each document by cosine similarity
    let mut scored_hits: Vec<(AnswerHit, f32, usize, usize)> = Vec::with_capacity(tfidf_index.documents.len());
    for (document_index, document_entry) in tfidf_index.documents.iter().enumerate() {
        let document_norm = tfidf_index.document_l2_norms[document_index];
        if document_norm == 0.0 {
            continue;
        }

        // Dot product only over query terms present in the document
        let mut dot_product_sum: f32 = 0.0;
        for (term_text, query_weight) in &query_weighted_map {
            if let Some(document_tf) = document_entry.term_frequencies.get(term_text) {
                let idf_value = *tfidf_index
                    .inverse_document_frequency
                    .get(term_text)
                    .unwrap_or(&0.0);
                if idf_value != 0.0 {
                    let document_weight = (*document_tf) * idf_value;
                    dot_product_sum += query_weight * document_weight;
                }
            }
        }

        if dot_product_sum == 0.0 {
            continue;
        }

        let cosine_similarity = dot_product_sum / (query_l2_norm * document_norm);

        // Tie-breakers need overlap count and question length
        let document_token_set = document_entry
            .term_frequencies
            .keys()
            .cloned()
            .collect::<HashSet<String>>();
        let overlap_count = document_token_set
            .intersection(&query_token_set)
            .count();

        let answer_hit = make_hit(
            document_entry.guid.clone(),
            document_entry.deck_path.clone(),
            Some(document_entry.question_preview.clone()),
            cosine_similarity,
        );

        scored_hits.push((
            answer_hit,
            cosine_similarity,
            overlap_count,
            document_entry.token_count,
        ));
    }

    // 4) Sort with tie-breakers: higher similarity, then higher overlap, then shorter question, then lexicographic GUID
    scored_hits.sort_by(|left, right| {
        right.1.partial_cmp(&left.1).unwrap_or(std::cmp::Ordering::Equal)
            .then_with(|| right.2.cmp(&left.2)) // greater overlap wins
            .then_with(|| left.3.cmp(&right.3)) // shorter question wins
            .then_with(|| left.0.guid.cmp(&right.0.guid))
    });

    scored_hits
        .into_iter()
        .take(top_k)
        .map(|tuple| tuple.0)
        .collect()
}
